pyhf/__init__.py

```diff
@@ -307,53 +308,68 @@ def generate_asimov_data(asimov_mu, data, pdf, init_pars, par_bounds):
+        pdf (Tensor): The model used in the likelihood ratio calculation
+        init_pars (Tensor): The initial parameters
+        par_bounds(Tensor): The bounds on the paramter values
+
+    Returns:
+        Float: The calculated test statistic, q_mu
+    """
+    mubhathat = optimizer.constrained_bestfit(
+        loglambdav, mu, data, pdf, init_pars, par_bounds)
+    muhatbhat = optimizer.unconstrained_bestfit(
+        loglambdav, data, pdf, init_pars, par_bounds)
+    qmu = loglambdav(mubhathat, data, pdf) - loglambdav(muhatbhat, data, pdf)
+    qmu = tensorlib.where(muhatbhat[pdf.config.poi_index] > mu, [0], qmu)
     return qmu
 
-from scipy.stats import norm
+
 def pvals_from_teststat(sqrtqmu_v, sqrtqmuA_v):
     # these pvals are from formula
     # (59) in arxiv:1007.1727 p_mu = 1-F(q_mu|mu') = 1- \Phi(q_mu - (mu-mu')/sigma)
     # and  (mu-mu')/sigma = sqrt(Lambda)= sqrt(q_mu_A)
-    CLsb = 1 - norm.cdf(sqrtqmu_v)
-    CLb =  1 - norm.cdf(sqrtqmu_v - sqrtqmuA_v)
+    CLsb = 1 - tensorlib.normal_cdf(sqrtqmu_v)
+    CLb = 1 - tensorlib.normal_cdf(sqrtqmu_v - sqrtqmuA_v)
     oneOverCLs = CLb / CLsb
     return CLsb, CLb, oneOverCLs
 
-import math
+
 def runOnePoint(muTest, data, pdf, init_pars, par_bounds):
-    asimov_mu = 0.0
-    asimov_data = tensorlib.tolist(generate_asimov_data(asimov_mu, data,
-                                   pdf, init_pars, par_bounds))
-
-    qmu_v  = qmu(muTest, data, pdf, init_pars, par_bounds)
-    qmuA_v = qmu(muTest, asimov_data, pdf, init_pars, par_bounds)
-
-    try:
-        sqrtqmu_v = math.sqrt(qmu_v)
-    except ValueError:
-        log.warning('WARNING: qmu_v negative: %s', qmu_v)
-        qmu_v = tensorlib.tolist(tensorlib.clip(qmu_v, 0, max=None))
-        sqrtqmu_v = math.sqrt(qmu_v)
-    try:
-        sqrtqmuA_v = math.sqrt(qmuA_v)
-    except ValueError:
-        log.warning('WARNING: qmuA_v negative: %s', qmuA_v)
-        qmuA_v = tensorlib.tolist(tensorlib.clip(qmuA_v, 0, max=None))
-        sqrtqmuA_v = math.sqrt(qmuA_v)
-
-    sigma = muTest / sqrtqmuA_v if sqrtqmuA_v > 0 else None
+    asimov_mu = 0.
+    asimov_data = generate_asimov_data(
+        asimov_mu, data, pdf, init_pars, par_bounds)
+
+    qmu_v = tensorlib.clip(
+        qmu(muTest, data, pdf, init_pars, par_bounds), 0, max=None)
+    sqrtqmu_v = tensorlib.sqrt(qmu_v)
+
+    qmuA_v = tensorlib.clip(
+        qmu(muTest, asimov_data, pdf, init_pars, par_bounds), 0, max=None)
+    sqrtqmuA_v = tensorlib.sqrt(qmuA_v)
 
     CLsb, CLb, oneOverCLs = pvals_from_teststat(sqrtqmu_v, sqrtqmuA_v)
 
     oneOverCLs_exp = []
     for nsigma in [-2, -1, 0, 1, 2]:
         sqrtqmu_v_sigma = sqrtqmuA_v - nsigma
-        oneOverCLs_exp.append(pvals_from_teststat(sqrtqmu_v_sigma, sqrtqmuA_v)[-1])
-    return qmu_v, qmuA_v, sigma, CLsb, CLb, oneOverCLs, oneOverCLs_exp
+        oneOverCLs_exp.append(
+            pvals_from_teststat(sqrtqmu_v_sigma, sqrtqmuA_v)[-1])
+    return qmu_v, qmuA_v, CLsb, CLb, oneOverCLs, oneOverCLs_exp
```

Comment:
thinking whether we should wrap `tensorlib.astensor(oneOverCLs_exp)`, this will make the expected CLs values a (5,1) tensor, which is a bit nicer to handle than 5 (1,) tensors

Reply:
Sounds good to me.

Reply:
Fixed in commit 0691e21